src/traffic_server/traffic_server.cc

```diff
@@ -224,7 +224,12 @@ struct AutoStopCont : public Continuation {
   {
     APIHook *hook = lifecycle_hooks->get(TS_LIFECYCLE_SHUTDOWN_HOOK);
     while (hook) {
-      hook->invoke(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
+      if (hook->m_cont->mutex) {
```

Comment:
Why not
```
SCOPED_MUTEX_LOCK(lock, hook->m_cont->mutex, this_ethread());
hook->invokle(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
```
?

Reply:
So the lock object will be a harmless do-nothing if the mutex is null?

Reply:
Yes. It was designed that way precisely to get this effect. Better to check the null in one place, and not every callsite.

Reply:
To be OCD about not holding mutexes longer than necessary, I would prefer:
```
while (hook) {
  {
    SCOPED_MUTEX_LOCK(lock, hook->m_cont->mutex, this_ethread());
    hook->invoke(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
  }
  hook = hook->next;
}
```

Reply:
It's during shutdown, I think I want it to hold it for longer, might expose other problems in the code.